### DISCUSSION TITLE
PATHNAME in the report name is blank + assertions aren't be run on all the urls

### DISCUSSION THREAD

**Discussion by jaspritt-hora** [GoogleChrome/lighthouse-ci#834] [Q&A]
Hello there, 
I am generating lighthouse reports for my app with a local setup and the reports are dumped on the `filesystem`. 

I have two questions -

1.  When running `lhci autorun` for multiple urls of the same website, the reports doesn't have the `PATHNAME` (it is blank) even though it is passed in `reportFilenamePattern` option.
2. The assertions specified is run on the `base url` instead on the `provided url`?

**Additional context for Q1**: 

`reportFilenamePattern` has the format : `%%HOSTNAME%%-%%PATHNAME%%-%%DATETIME%%-report.%%EXTENSION%%` 

_Current behaviour_: 

<img width="364" alt="image" src="https://user-images.githubusercontent.com/20161217/196516908-9d1bf454-7d2c-42ba-ba8a-6d85dcde6015.png">

_Expected behaviour_: 
The report name should have the following - 

`localhost-users-<date time>.json/.html`
`localhost-messages-<date time>.json/.html`
`localhost-calendar-<date time>.json/.html`

What else I am missing here?

**Additional context for Q2**: 

The `urls` option is passed the following values: 
`url:[
  'http://localhost:8070/#/en/',
  'http://localhost:8070/#/en/users/',
  'http://localhost:8070/#/en/calendar/',
  'http://localhost:8070/#/en/library/?tab=exercises',
  'http://localhost:8070/#/en/facility_assignments/',
  'http://localhost:8070/#/en/messages/',
  'http://localhost:8070/#/en/settings?tab=risk'
] `
                
_Output of the `autorun` command (Current behaviour)_: 

It says - 

`Checking assertions against 1 URL(s), 7 total run(s)`

<img width="602" alt="image" src="https://user-images.githubusercontent.com/20161217/196517213-4c8fab55-f36e-42ec-9e97-12fd9cd2900c.png">

_Expected behaviour_: 

It should say - 

`Checking assertions against 7 URL(s), 1 total run(s)` since there are 7 separate urls

**Overall information and details**: 

`lighthouserc.js`

<img width="1026" alt="image" src="https://user-images.githubusercontent.com/20161217/196519124-6450716e-b81b-4abe-ba9d-3a85d24c2953.png">

lhci - `v0.9.0`
Node - `v16`

**Comment by jaspritt-hora** [comment]
@patrickhulce - Can you let me know, what needs to be done here?

**Reply by patrickhulce** [reply]
Everything here is WAI because the URLs here have no unique pathname, they use the client-side routing hack of `#/` so this would require a new feature to support that in the template.

**Reply by jaspritt-hora** [reply]
@patrickhulce - Thank you for the info. Is there any workaround for this currently?
Additionally, do you have an idea about when this change would be taken up for implementation?

**Reply by patrickhulce** [reply]
>Is there any workaround for this currently?

Not aside from using URLs with an actual pathname. 

> do you have an idea about when this change would be taken up for implementation?

As soon as you have some free time ;) (just need to add `url.hash` to https://github.com/GoogleChrome/lighthouse-ci/blob/75f4bb8f66d5042989f3882683c0e843677dda06/packages/cli/src/upload/upload.js#L545)

**Reply by jaspritt-hora** [reply]
@patrickhulce - Sounds good. I can take up the addition. This is going to be the first time I am contributing to open source; anything apart from the contributing guidelines needed to be kept in mind before working on this feature?

**Comment by jaspritt-hora** [comment]
@patrickhulce - Can you please create a new release & tag with [this](https://github.com/GoogleChrome/lighthouse-ci/pull/835) change, so that it can be used in the app?

**Reply by patrickhulce** [reply]
I've recently had a baby and am limiting OSS involvement. Perhaps @paulirish or @connorjclark could help out here?

**Reply by jaspritt-hora** [reply]
@paulirish or @connorjclark - Can you please help out here in regards to the release?

**Reply by jaspritt-hora** [reply]
@connorjclark - Thank you for the release for the above implementation.

I have updated the `lhci` in my app to the latest version i.e. `0.10.0` In which I have added `url hash` for the file upload report name. 
In the `upload` section of `lighthouserc.js` - 

The `reportFilenamePattern` is set to - `%%HOSTNAME%%-%%PATHNAME%%-%%HASH%%-%%DATETIME%%.report.%%EXTENSION%%`

`hash` is the new parameter that I have added as part of the PR.  The test case for the same is written in `upload-url-hash.test.js`. It is working as expected and contains 3 urls with `hash`.
 Refer screenshot below for implementation -

<img width="426" alt="image" src="https://user-images.githubusercontent.com/20161217/203583079-91b35db6-1e52-4199-b8ff-7e2e94bfb196.png">

Screenshot of the test case with url hash -
<img width="872" alt="image" src="https://user-images.githubusercontent.com/20161217/203598874-2c2703c0-40fa-416b-9013-f51b39663c39.png">


After making the above changes in my app to `lighthouserc.js` file, even after running `lhci autorun` command, the `hash` parameter is printed as `unknown`

The url against which I am running lighthouse is 
`https://www.example.com/#/page3`

<img width="493" alt="image" src="https://user-images.githubusercontent.com/20161217/203582468-d41e1eb3-36be-4738-ae48-695f4896284f.png">

Instead of `unknown`, it should have been `page3`.
Additionally, in the `node modules` installed for my app - has the updated version and the url hash implementation. 

**Question** - What am I missing here?

**Reply by jaspritt-hora** [reply]
@connorjclark / @patrickhulce - Can you please help out here?

**Reply by connorjclark** [reply]
The reason comes down to `lhr.finalUrl` never containing hashes, because it is a network url and hashes are client only. see comments here:  https://github.com/GoogleChrome/lighthouse/blob/bb2bff7f5fd665919adefac2f55e05af9b3bba43/types/lhr/lhr.d.ts#L24

We should change the hash to grab from the `lhr.requestedUrl`. Can you send a PR for that?

For >= 10.0 (unreleased), we'd instead use `finalDisplayedUrl`. But, it's made me wonder if these filenames should use the _requested_ url state instead of the final, resolved url. Any thought on that?

**Reply by jaspritt-hora** [reply]
@connorjclark - 

> The reason comes down to `lhr.finalUrl` never containing hashes, because it is a network url and hashes are client only. see comments here: https://github.com/GoogleChrome/lighthouse/blob/bb2bff7f5fd665919adefac2f55e05af9b3bba43/types/lhr/lhr.d.ts#L24

I see. That explains the scenario. 

> We should change the hash to grab from the `lhr.requestedUrl`. Can you send a PR for that?

I can create a PR for the same. However, I have follow-up questions - 
- The change that we are talking about covers the cases in which the reports are generated through the `filesystem` option. What's your take on the other cases i.e. uploading to `lhci server` and using `temporary storage`?

I believe the `hash` isn't being considered there as well. That being said, this would then increase the changes to multiple aspects of the functionality.

Refer [similar discussion for uploading to lhci server](https://github.com/GoogleChrome/lighthouse-ci/discussions/832)
- `Reasoning` - The `upload` behaviour would be disjointed ,if the app is using url hashes in cases that do not involve `filesystem`.

> For >= 10.0 (unreleased), we'd instead use `finalDisplayedUrl`. But, it's made me wonder if these filenames should use the _requested_ url state instead of the final, resolved url. Any thought on that?

In theory and logically , it makes sense to use the `requestedUrl` to generate filenames. However, that leads me to a more important question - what is/was the reasoning to use `lhr.finalUrl` to generate filenames?

Additionally, a quick search in code for `lhr.finalUrl` leads me to believe that this would require a lot testing and would need your assistance in the same :)

**Reply by jaspritt-hora** [reply]
@connorjclark - Here is the PR #849 for the `requestedUrl` change for `filesystem` upload only.

**Reply by jaspritt-hora** [reply]
@connorjclark - Could you please spare a few minutes and review the PR and answer a few followup questions, in the thread above? 
Thank you!
Have a good one!

**Reply by jaspritt-hora** [reply]
@connorjclark / @patrickhulce - Any updates on the above thread and the PR?

**Reply by patrickhulce** [reply]
LGTM but I'll defer to @connorjclark for an approval / merge / release :)

**Reply by jaspritt-hora** [reply]
What do you think about doing a similar change for the other options i.e. `lhci server` and `temporary storage`?. The hash isn't handled there as well.
Additionally, what was the reasoning behind referencing `finalUrl` instead of `requestedUrl` for uploading reports in the first place?

**Reply by jaspritt-hora** [reply]
@connorjclark - Could you please review the PR and answer the follow up questions above?

**Reply by jaspritt-hora** [reply]
@connorjclark - Request you to  please review the PR and provide your approvals or your comments and complete the other steps in the workflow. So, that we can test this out in our application - as we are actively working on the integration. 

Thank you!

**Reply by connorjclark** [reply]
@jaspritt-hora I suggest you figure out how to utilize your fork more directly. Given the holidays this isn't going to happen on your expected timeline.

**Reply by jaspritt-hora** [reply]
@connorjclark - Thank you for the update. Will use the fork and test out the change.
Happy holidays!

**Reply by jaspritt-hora** [reply]
@connorjclark - I have tested my fork and the PR raised for `url hash` is working flawlessly for `filesystem`.  Can you tell me a ballpark ETA for the new release that would include the PR?
